cmd/ttn-lw-cli/commands/end_devices.go

```diff
@@ -949,22 +949,87 @@ var (
 			logger.WithField("paths", nsPaths).Debug("Reset end device to factory defaults on Network Server")
 			nsDevice, err := ttnpb.NewNsEndDeviceRegistryClient(ns).ResetFactoryDefaults(ctx, &ttnpb.ResetAndGetEndDeviceRequest{
 				EndDeviceIds: devID,
-				FieldMask:    ttnpb.FieldMask(nsPaths...),
+				FieldMask:    ttnpb.FieldMask(ttnpb.AddFields(nsPaths, "supports_join")...),
+			})
+			if err != nil {
+				return err
+			}
+			if err = device.SetFields(nsDevice, "ids.dev_addr"); err != nil {
+				return err
+			}
+			if err = device.SetFields(nsDevice, ttnpb.AllowedBottomLevelFields(nsPaths, getEndDeviceFromNS)...); err != nil {
+				return err
+			}
+			device.UpdateTimestamps(nsDevice)
+
+			as, err := api.Dial(ctx, config.ApplicationServerGRPCAddress)
+			if err != nil {
+				return err
+			}
+			logger.WithField("paths", asPaths).Debug("Reset end device to factory defaults on Application Server")
+			asDevice, err := ttnpb.NewAsEndDeviceRegistryClient(as).Get(ctx, &ttnpb.GetEndDeviceRequest{
```

Comment:
Why is the `Get` operation needed here ? If you provide the field masks with an empty source, you are in effect unsetting that particular value.
There is no need for `Get` -> `SetFields` -> `Set`, you can do just the `Set`.